dspace-api/src/main/java/org/dspace/app/bulkdo/ActionTarget.java

```diff
@@ -0,0 +1,284 @@
+import org.dspace.authorize.service.ResourcePolicyService;
+import org.dspace.content.*;
+import org.dspace.content.factory.ContentServiceFactory;
+import org.dspace.content.service.ItemService;
+import org.dspace.core.Constants;
+import org.dspace.core.Context;
+
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+
+class ActionTarget {
+
+    protected Context context;
+    private DSpaceObject obj;
+    private HashMap<String, Object> map;
+    private ActionTarget up;
+
+    public static final String POLICY = "POLICY";
+
+    ActionTarget(Context contxt, ActionTarget container, DSpaceObject o) {
+        assert (o != null);
+        context = contxt;
+        obj = o;
+        up = container;
+        map = null;
+    }
+
+    public DSpaceObject getObject() {
+        return obj;
+    }
+
+    public static ActionTarget create(Context context, ActionTarget container, DSpaceObject obj) {
+        assert (obj != null);
+        switch (obj.getType()) {
+            case Constants.BITSTREAM:
+                return new BitstreamActionTarget(context, container, obj);
+            case Constants.BUNDLE:
+                return new BundleActionTarget(context, container, obj);
+            case Constants.ITEM:
+                return new ItemActionTarget(context, container, obj);
+            case Constants.COLLECTION:
+                return new CollectionActionTarget(context, container, obj);
+            case Constants.COMMUNITY:
+                return new ActionTarget(context, container, obj);
+            default:
+                assert (false);
+                throw new RuntimeException("should never try to create ActionTarget from " + obj);
+        }
+    }
+
+    public static ActionTarget createUpFor(Context context, DSpaceObject obj) {
+        assert (obj != null);
+        try {
+            DSpaceObject up = null;
+            switch (obj.getType()) {
+                case Constants.BITSTREAM:
+                    up = ((Bitstream) obj).getBundles().get(0);
+                    break;
+                case Constants.BUNDLE:
+                    up = ((Bundle) obj).getItems().get(0);
+                    break;
+                case Constants.ITEM:
+                case Constants.COLLECTION:
+                    up = ContentServiceFactory.getInstance().getDSpaceObjectService(obj).getParentObject(context,obj);
+                    break;
+                case Constants.COMMUNITY:
+                    up = null;
+                    break;
+                default:
+                    throw new RuntimeException("should never try to create ActionTarget from " + obj.toString());
+            }
+            if (up == null)
+                return null;
+            return create(context, null, up);
+        } catch (SQLException e) {
+            throw new RuntimeException("should never happen???:  " + e.getMessage());
```

Comment:
Best to log the exception, it could always happen so perhaps change the message ?